core/constraint-streams/src/main/java/ai/timefold/solver/constraint/streams/bavet/common/AbstractConcatNode.java

```diff
@@ -0,0 +1,103 @@
+package ai.timefold.solver.constraint.streams.bavet.common;
+
+import static ai.timefold.solver.constraint.streams.bavet.common.tuple.TupleState.ABORTING;
+import static ai.timefold.solver.constraint.streams.bavet.common.tuple.TupleState.CREATING;
+import static ai.timefold.solver.constraint.streams.bavet.common.tuple.TupleState.DYING;
+
+import ai.timefold.solver.constraint.streams.bavet.common.tuple.AbstractTuple;
+import ai.timefold.solver.constraint.streams.bavet.common.tuple.LeftTupleLifecycle;
+import ai.timefold.solver.constraint.streams.bavet.common.tuple.RightTupleLifecycle;
+import ai.timefold.solver.constraint.streams.bavet.common.tuple.TupleLifecycle;
+import ai.timefold.solver.constraint.streams.bavet.common.tuple.TupleState;
+
+public abstract class AbstractConcatNode<Tuple_ extends AbstractTuple>
+        extends AbstractNode
+        implements LeftTupleLifecycle<Tuple_>, RightTupleLifecycle<Tuple_> {
+    private final int inputStoreIndexLeftOutTupleList;
+    private final int inputStoreIndexRightOutTupleList;
+    protected final int outputStoreSize;
+    private final StaticPropagationQueue<Tuple_> propagationQueue;
+
+    protected AbstractConcatNode(TupleLifecycle<Tuple_> nextNodesTupleLifecycle,
+            int inputStoreIndexLeftOutTupleList,
+            int inputStoreIndexRightOutTupleList,
+            int outputStoreSize) {
+        this.propagationQueue = new StaticPropagationQueue<>(nextNodesTupleLifecycle);
+        this.inputStoreIndexLeftOutTupleList = inputStoreIndexLeftOutTupleList;
+        this.inputStoreIndexRightOutTupleList = inputStoreIndexRightOutTupleList;
+        this.outputStoreSize = outputStoreSize;
+    }
+
+    protected abstract Tuple_ getOutTuple(Tuple_ inTuple);
+
+    @Override
+    public final void insertLeft(Tuple_ tuple) {
+        Tuple_ outTuple = getOutTuple(tuple);
+        tuple.setStore(inputStoreIndexLeftOutTupleList, outTuple);
```

Comment:
If you store the node-specific tuple state in the store instead of the outtuple, you won't need to create the outtuple just to store the state.